File: python/task5.py
import math

def spliting(Message):
    return Message.split()

def Ascii(decimal_number):
    return chr(decimal_number)

def decode(Number):
    Array = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
             'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
             'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
             'U', 'V', 'W', 'X', 'Y', 'Z']
    output =  0

    for i in Number:
        for j in range(0, len(Array)):
            if i == Array[j]:
                output = output * len(Array) + j
    return output

def remainder(base, Number):
    numerical = decode(str(Number))  # Corrected to ensure decode receives a string
    wholeNumber = numerical // base
    remainder = numerical % base
    return (remainder, wholeNumber)

def index(character):
    Array = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
             'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
             'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
             'U', 'V', 'W', 'X', 'Y', 'Z']
    index = -1
    for i in range(0, len(Array)):
        if character == Array[i]:
            index = i
            break
    return index

def Decrypt(base, Number):
    Array = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
             'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
             'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
             'U', 'V', 'W', 'X', 'Y', 'Z']

    output = ""
    number = Number
    while number != 0:
        output = Array[remainder(base, number)[0]] + output
        number = int(remainder(base, number)[1])
    return output

def Encryption(base, Number):
    Array = ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
             'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
             'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
             'U', 'V', 'W', 'X', 'Y', 'Z']

    output = ""
    number = Number
    while number != 0:
        output = Array[int(remainder(base, number)[0])] + output
        number = int(remainder(base, number)[1])
    return output

def Decryption(base, Cipher_Text):
    Array = spliting(Cipher_Text)
    output = ""
    for i in Array:
        output += Ascii(int(Decrypt(base, i)))  # Corrected to convert result to integer
    return output

message = str(input())
baseinput = int(input())
userinput = str(input())

if message == "decrypt":
    print(Decryption(baseinput, userinput))
else:
    print(Encryption(baseinput, userinput))

File: c++/task5.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>

using namespace std;
using Array = vector<char>;
using Arrays = vector<string>;

pair<int, int> remainder(int base, int Number) {
    int wholeNumber = Number / base;
    int remainder = Number % base;
    return make_pair(remainder, wholeNumber);
}

string tobase(int decimal_num, int base) {
    string result = "";

    while (decimal_num > 0) {
        int remainder = decimal_num % base;
        if (remainder < 10) {
            result = to_string(remainder) + result;
        } else {
            result = static_cast<char>('A' + remainder - 10) + result;
        }
        decimal_num /= base;
    }

    return result;
}

void Encryption(int base, string userinput) {
    for (char c : userinput) {
        cout << tobase(static_cast<int>(c), base) << endl;
    }
}

int index(char character) {
    Array Array{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
                'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
                'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                'U', 'V', 'W', 'X', 'Y', 'Z'};

    for (int i = 0; i < Array.size(); ++i) {
        if (character == Array[i]) {
            return i;
        }
    }
    return -1;
}

int Encrypt(int base, string Number) {
    int output = 0;
    for (int i = 0; i < Number.length(); ++i) {
        output += index(Number[i]) * pow(base, i);
    }
    return output;
}

string Decrypt(int base, int Number) {
    Array Array{'0', '1', '2', '3', '4', '5', '6', '7', '8', '9',
                'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J',
                'K', 'L', 'M', 'N', 'O', 'P', 'Q', 'R', 'S', 'T',
                'U', 'V', 'W', 'X', 'Y', 'Z'};

    string output = "";
    int number = Number;
    while (number != 0) {
        output += Array[remainder(base, number).first];
        number = remainder(base, number).second;
    }
    reverse(output.begin(), output.end());
    return output;
}

int fromBase(string input_str, int base) {
    int value = 0;
    for (char digit : input_str) {
        value = base * value + stoi(string(1, digit), nullptr, base);
    }
    return value;
}

void Decryption(int base, Arrays userinput) {
    Array DecryptedArray;
    for (string input_str : userinput) {
        if (input_str == "-1") {
            break;
        }
        string decrypted_str = "";

        for (int i = 0; i < input_str.length(); i += 2) {
            decrypted_str += static_cast<char>(fromBase(input_str.substr(i, 2), base));
        }
        DecryptedArray.push_back(decrypted_str);
    }

    string output_str = "";
    for (string i : DecryptedArray) {
        output_str += i;
    }
    cout << output_str << endl;
}

int main() {
    string instruction;
    int baseinput;
    Arrays userinput;

    cin >> instruction;
    cin >> baseinput;

    if (instruction == "encrypt") {
        string input_str;
        while (cin >> input_str && input_str != "-1") {
            Encryption(baseinput, input_str);
        }
    } else {
        string input_str;
        while (cin >> input_str && input_str != "-1") {
            userinput.push_back(input_str);
        }
        Decryption(baseinput, userinput);
    }

    return 0;
}

